### PR TITLE
Rollup functionality.

### PR DESCRIPTION
**Pull request by derekcollison**
Messages published with the rollup header will place the new message onto the stream and purge all others based on subject, or the complete stream.

Signed-off-by: Derek Collison <derek@nats.io>

/cc @nats-io/core

### CODE REVIEW

**Review by ripienaar — APPROVED** [review]
very cool, LGTM

File: server/stream.go [hunk 1]
```diff
@@ -2931,6 +2960,11 @@ func (mset *stream) processJetStreamMsg(subject, reply string, hdr, msg []byte,
 		if msgId != _EMPTY_ {
 			mset.storeMsgId(&ddentry{msgId, seq, ts})
 		}
+		if rollupSub {
+			mset.purge(&JSApiStreamPurgeRequest{Subject: subject, Keep: 1})
+		} else if rollupAll {
+			mset.purge(&JSApiStreamPurgeRequest{Keep: 1})
```
Comment (ripienaar) [inline server/stream.go:2966]: should we error for rollup subject with unsupported value?

Comment (derekcollison) [reply server/stream.go:2966]: We do above.

Comment (ripienaar) [reply server/stream.go:2966]: doh, missed that!

**Comment by tpihl** [comment]
Looks very interesting

**Pull request merged**